airbyte-integrations/connectors/source-vertica/source_vertica/source.py

```diff
@@ -0,0 +1,173 @@
+from typing import Dict, Generator
+
+from airbyte_cdk.logger import AirbyteLogger
+from airbyte_cdk.models import (
+    AirbyteCatalog,
+    AirbyteConnectionStatus,
+    AirbyteMessage,
+    AirbyteRecordMessage,
+    AirbyteStream,
+    ConfiguredAirbyteCatalog,
+    Status,
+    Type,
+)
+from airbyte_cdk.sources import Source
+
+import vertica_python
+
+
+class SourceVertica(Source):
+    def check(self, logger: AirbyteLogger, config: json) -> AirbyteConnectionStatus:
+        """
+        Tests if the input configuration can be used to successfully connect to the integration
+            e.g: if a provided Stripe API token can be used to connect to the Stripe API.
+
+        :param logger: Logging object to display debug/info/error to the logs
+            (logs will not be accessible via airbyte UI if they are not passed to this logger)
+        :param config: Json object containing the configuration of this source, content of this json is as specified in
+        the properties of the spec.yaml file
+
+        :return: AirbyteConnectionStatus indicating a Success or Failure
+        """
+        try:
+            self.create_connection(config)
+
+            return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+        except Exception as e:
+            return AirbyteConnectionStatus(status=Status.FAILED, message=f"An exception occurred: {str(e)}")
+
+    def discover(self, logger: AirbyteLogger, config: json) -> AirbyteCatalog:
+        """
+        Returns an AirbyteCatalog representing the available streams and fields in this integration.
+        For example, given valid credentials to a Postgres database,
+        returns an Airbyte catalog where each postgres table is a stream, and each table column is a field.
+
+        :param logger: Logging object to display debug/info/error to the logs
+            (logs will not be accessible via airbyte UI if they are not passed to this logger)
+        :param config: Json object containing the configuration of this source, content of this json is as specified in
+        the properties of the spec.yaml file
+
+        :return: AirbyteCatalog is an object describing a list of all available streams in this source.
+            A stream is an AirbyteStream object that includes:
+            - its stream name (or table name in the case of Postgres)
+            - json_schema providing the specifications of expected schema for this stream (a list of columns described
+            by their names and types)
+        """
+        streams = []
+
+        connection = self.create_connection(config)
+
+        tables = self.get_tables(logger, config, connection)
+
+        for table in tables:
+            properties = {}
+            columns = self.get_columns(logger, config, connection, table)
+
+            for column in columns:
+                properties[column[0]] = {
+                    "type": column[1]
+                }
+
+            streams.append(
+                AirbyteStream(
```

Comment:
Couple of things here:
1. What are the supported sync modes that you are planning to support? If you plan to support incremental syncs you will need to implement some state management in order to store the previously synced date.
2. You can define a source defined primary key which is the db primary key which can uniquely identify each row.
3. I don't think you can directly map the column name/type properties to a schema (https://json-schema.org/). You will most likely need to do some type translation between the db types and the supported Airbyte types (https://docs.airbyte.com/understanding-airbyte/supported-data-types/)

Reply:
For 1 and 2, could you maybe give me an example where this has been achieved?
For 3, yes, I will do this mapping.

Reply:
Hmm I don't know if there are any db examples implemented in Python (that is why I have suggested the jdbc implementation). Maybe look into the internals of the python cdk on how it transmits state when reading from an http source? If you don't think you can make it for the deadline maybe you can only support full refresh for the initial release.

Reply:
@itaseskii Yes, that would be great. I will aim to understand how an incremental sync works and expand upon it later.